base_transport.go

```diff
@@ -0,0 +1,108 @@
+package rollbar
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+type baseTransport struct {
+	// Rollbar access token used by this transport for communication with the Rollbar API.
+	Token string
+	// Endpoint to post items to.
+	Endpoint string
+	// Logger used to report errors when sending data to Rollbar, e.g.
+	// when the Rollbar API returns 409 Too Many Requests response.
+	// If not set, the client will use the standard log.Printf by default.
+	Logger ClientLogger
+	// RetryAttempts is how often to attempt to resend an item when a temporary network error occurs
+	// This defaults to DefaultRetryAttempts
+	// Set this value to 0 if you do not want retries to happen
+	RetryAttempts int
+	// PrintPayloadOnError is whether or not to output the payload to the set logger or to stderr if
+	// an error occurs during transport to the Rollbar API.
+	PrintPayloadOnError bool
+	// custom http client (http.DefaultClient used by default)
+	httpClient *http.Client
+}
+
+// SetToken updates the token to use for future API requests.
+func (t *baseTransport) SetToken(token string) {
+	t.Token = token
+}
+
+// SetEndpoint updates the API endpoint to send items to.
+func (t *baseTransport) SetEndpoint(endpoint string) {
+	t.Endpoint = endpoint
+}
+
+// SetLogger updates the logger that this transport uses for reporting errors that occur while
+// processing items.
+func (t *baseTransport) SetLogger(logger ClientLogger) {
+	t.Logger = logger
+}
+
+// SetRetryAttempts is how often to attempt to resend an item when a temporary network error occurs
+// This defaults to DefaultRetryAttempts
+// Set this value to 0 if you do not want retries to happen
+func (t *baseTransport) SetRetryAttempts(retryAttempts int) {
+	t.RetryAttempts = retryAttempts
+}
+
+// SetPrintPayloadOnError is whether or not to output the payload to stderr if an error occurs during
+// transport to the Rollbar API.
+func (t *baseTransport) SetPrintPayloadOnError(printPayloadOnError bool) {
+	t.PrintPayloadOnError = printPayloadOnError
+}
+
+// SetHTTPClient sets custom http client. http.DefaultClient is used by default
+func (t *baseTransport) SetHTTPClient(c *http.Client) {
```

Comment:
I'm pretty sure this needs to be protected by a mutex otherwise it's a race condition once the transport is actively sending things to rollbar.

Reply:
PS: I believe that is also true of all of the above methods as well.

Reply:
Well, technically you're right, but in reality nothing bad can happen - pointers aren't spread through multiple machine words (at least on most architectures).

Reply:
There seem to be a variety of potential concurrency issues. I'd rather address them together, and not burden this PR with it.